src/inmanta/agent/handler.py

```diff
@@ -434,16 +435,146 @@ class ResourceHandler(object):
+
+        resources_in_unexpected_state = filter_resources_in_unexpected_state(requires)
+        if resources_in_unexpected_state:
+            ctx.set_status(const.ResourceState.skipped)
+            ctx.warning(
+                "Resource %(resource)s skipped because a dependency is in an unexpected state: %(unexpected_states)s",
+                resource=resource.id.resource_version_str(),
+                unexpected_states=str({rid: state.value for rid, state in resources_in_unexpected_state.items()}),
+            )
+            return
+
+        failed_dependencies = [req for req, status in requires.items() if status != ResourceState.deployed]
+        if not any(failed_dependencies):
+            self.execute(ctx, resource)
+            if _should_reload():
+                self.do_reload(ctx, resource)
+        else:
+            ctx.set_status(const.ResourceState.skipped)
+            ctx.info(
+                "Resource %(resource)s skipped due to failed dependencies: %(failed)s",
+                resource=resource.id.resource_version_str(),
+                failed=str(failed_dependencies),
+            )
+
+    @abstractmethod
+    def execute(self, ctx: HandlerContext, resource: resources.Resource, dry_run: bool = False) -> None:
+        """
+        This method contains the logic to enforce a resource's intent.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to deploy.
+        :param dry_run: If set to true, the intent is not enforced, only the set of changes it would bring is computed.
+        """
+        pass
+
+    @abstractmethod
+    def check_facts(self, ctx: HandlerContext, resource: resources.Resource) -> Dict[str, object]:
+        """
+        This method is called by the agent to query for facts.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to query facts for.
+        :return: A dict with fact names as keys and facts values.
+        """
+        pass
+
+    def available(self, resource: resources.Resource) -> bool:
+        """
+        Override this method to implement custom handler selection logic, e.g. if multiple handlers were defined
+        for the same resource type.
+
+        :param resource: Resource for which to check whether this handler is available.
+        """
+        return True
+
+    def can_reload(self) -> bool:
+        """
+        Can this handler reload?
+
+        :return: Return true if this handler needs to reload on requires changes.
+        """
+        return False
+
+    def do_reload(self, ctx: HandlerContext, resource: resources.Resource) -> None:
+        """
+        Perform a reload of this resource.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to reload.
+        """
```

Comment:
Had to lift these two as well since they're called from the `deploy` method. Suggestions for better docstrings are welcome

Reply:
Why lift the implementation of deploy? 

My understanding is that this class is purely the definition of the interface between agent and handler, so the deploy logic is not part of that?

Reply:
Most of what it does is relevant for the generic concept of a resource, isn't it? We could choose to extract some parts of it to the `Resource` subclass, but any resource has dependencies and its handler should generally be executed when those dependencies have successfully deployed.

Reply:
As discussed: keep it as is, but unmark HandlerABC stable_api annotation